### PR TITLE
fix(windows): BCP 47 tag canonicalization

### LINKED ISSUE #3485: [Windows] NW Iranian Latin Keyboard installed for Arabic instead of Latin script
# Original report

Something changed in Keyman fairly recently such that my keyboard for NW Iranian languages (https://keyman.com/keyboards/nw_iranian_latin) now tries to make Windows that it is a RTL script. I suppose it has something to do with the languages assigned to it which are normally Arabic script languages. But this is for writing in Latin script.

What do I need to change to get it to be recognized as a LTR Latin keyboard?

# Lorna's reply

The nw Iranian Keyboard is a Latin keyboard that has recently changed to RTL on one person's computer at least.

Do you know if this is a bug in Keyman (or a known bug)? I doublechecked and the .kmn keyboard does NOT say it's RTL. I also double-checked the tag in the .kps and the keyboard_info and it says "glk-Latn".

Then I double-checked langtags and that script code (glk-Latn) is in langtags.

So, I feel it's something in Keyman, but I don't know what.

# Marc's reply
> The right-to-left setting for keyboards in Windows is controlled by Windows and is based on the language code. I'd check the language association in Keyman Configuration first -- it may be that Windows recognises 'glk' and treats it as RTL regardless of the script. If that is the case, we'll need to change the language code (and it would be worth reporting the issue to Microsoft through the feedback tool). But first step is to verify the language code associated with the keyboard.

# Follow up report
 
This is how it got installed when I ran the Keyman installer for NW Iranian:

image

image

So it looks like it got automatically associated with Gilaki  (Arabic). I’m not sure where Windows gets this and maybe it’s new. I tried a couple minority languages to ‘add’ to Language preferences and it finds them which is interesting.

Gilaki would by default be in Arabic script, but this keyboard is for writing various NW Iranian languages in Latin transcription.

### PR DESCRIPTION
**Pull request by mcdurdin**
Reworks the BCP 47 tag canonicalization algorithm.

`TCanonicalLanguageCodeUtils.FindBestTag` will canonicalize known tags, then apply rules to ensure script subtag is present if not suppressed, and add a default region if none given.

Because the algorithm depends on the langtags.json data, which is large, where possible we use kmcomapi implementation to avoid bulking up executables, and refactored the `TBCP47Tag` class to avoid a dependency on this.

Future work: turn all the language data into a data file that is included in the installer (or possibly a separate dll) so that Developer doesn't end up with multiple copies of the data also.

Fixes #3485.
Part of #3512.
Relates to #1719.
Relates to #1282.

Recommend reviewing with ignore-whitespace.

### CODE REVIEW

File: windows/src/desktop/kmshell/install/UfrmInstallKeyboardLanguage.pas [hunk 1]
```diff
@@ -409,34 +411,63 @@ procedure TfrmInstallKeyboardLanguage.FillLanguageGrid;
...
+          if not TLanguageCodeUtils.BCP47Regions.TryGetValue(Region, FCustomLanguage.RegionName) then
+            FCustomLanguage.RegionName := '';
+          FCustomLanguage.FullName := TLanguageCodeUtils.LanguageName(FCustomLanguage.LanguageName,
+            FCustomLanguage.ScriptName, '');
+          AddRow(
+            False,
+            FCustomLanguage.FullName,
+            FCustomLanguage.FullName,
+            FCustomLanguage.ScriptName,
+            FCustomLanguage.Tag, nil);
```
Comment (darcywong00) [inline windows/src/desktop/kmshell/install/UfrmInstallKeyboardLanguage.pas:462]: This block looks a little WET 😉  ll 420-441

File: windows/src/unit-tests/standards-data/Keyman.Test.System.CanonicalLanguageCodeUtilsTest.pas [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+implementation
+
+uses
+  Keyman.System.CanonicalLanguageCodeUtils;
+
+{ TCanonicalLanguageCodeUtilsTest }
+
+procedure TCanonicalLanguageCodeUtilsTest.TestSomeTags;
+begin
+  // #3485 - Gilaki (Latin) script
```
Comment (darcywong00) [inline windows/src/unit-tests/standards-data/Keyman.Test.System.CanonicalLanguageCodeUtilsTest.pas:25]: This tests both (Arab) and (Latin) scripts.

What about also testing this?
Assert.AreNotEqual('glk-Latin-IR', TCanonicalLanguageCodeUtils.FindBestTag('glk'));

**Review by darcywong00 — APPROVED** [review]
lgtm

File: windows/src/desktop/kmshell/install/UfrmInstallKeyboardLanguage.pas (hunk 1 above)
Comment (mcdurdin) [reply windows/src/desktop/kmshell/install/UfrmInstallKeyboardLanguage.pas:462]: Fixed in https://github.com/keymanapp/keyman/pull/3557/files#diff-3a8f936acbcb68d130f21fa38f1733e8L421

File: windows/src/unit-tests/standards-data/Keyman.Test.System.CanonicalLanguageCodeUtilsTest.pas (hunk 2 above)
Comment (mcdurdin) [reply windows/src/unit-tests/standards-data/Keyman.Test.System.CanonicalLanguageCodeUtilsTest.pas:25]: Given `Assert.AreEqual('glk-Arab-IR', TCanonicalLanguageCodeUtils.FindBestTag('glk'));` already passed, I don't think we need to try and test the inverse as well 😁

**Pull request merged**